src/main/java/rx/internal/operators/OnSubscribeRedo.java

```diff
@@ -235,10 +235,11 @@ public void onError(Throwable e) {
                     @Override
                     public void onNext(T v) {
                         if (!done) {
-                            if (consumerCapacity.get() != Long.MAX_VALUE) {
-                                consumerCapacity.decrementAndGet();
-                            }
                             child.onNext(v);
+                            long cap = consumerCapacity.get();
+                            if (cap != Long.MAX_VALUE) {
+                                consumerCapacity.compareAndSet(cap, cap -1);
```

Comment:
The problem with this is that if the CAS fails, the 'production' of the value is not accounted and the capacity seems to rise without limit. The options are 1) add a loop that makes sure the decrement actually happens or 2) switch back to decrementAndGet and don't worry about request getting to max concurrently.